packages/transaction-controller/src/utils/simulation.ts

```diff
@@ -0,0 +1,435 @@
+    const value = parseEventArgValue(arg);
+
+    result[name] = value;
+
+    return result;
+  }, {});
+}
+
+/**
+ * Parse an event argument value.
+ * @param value - The event argument value.
+ * @returns The parsed event argument value.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function parseEventArgValue(value: any): any {
+  if (Array.isArray(value)) {
+    return value.map(parseEventArgValue);
+  }
+
+  let parsedValue = value;
+
+  if (parsedValue.toHexString) {
+    parsedValue = value.toHexString();
+  }
+
+  return parsedValue.toLowerCase();
+}
+
+/**
+ * Generate token balance changes from parsed events.
+ * @param request - The transaction that was simulated.
+ * @param events - The parsed events.
+ * @returns An array of token balance changes.
+ */
+async function getTokenBalanceChanges(
+  request: GetSimulationDataRequest,
+  events: ParsedEvent[],
+): Promise<SimulationTokenBalanceChange[]> {
+  const balanceTransactionsByToken = getTokenBalanceTransactions(
+    request,
+    events,
+  );
+
+  const balanceTransactions = [...balanceTransactionsByToken.values()];
+
+  log('Generated balance transactions', balanceTransactions);
+
+  if (!balanceTransactions.length) {
+    return [];
+  }
+
+  const response = await simulateTransactions(request.chainId as Hex, {
+    transactions: [...balanceTransactions, request, ...balanceTransactions],
+  });
+
+  if (response.transactions.length !== balanceTransactions.length * 2 + 1) {
+    throw new Error('Invalid response from simulation API');
+  }
+
+  log('Balance simulation response', response);
+
+  return [...balanceTransactionsByToken.keys()]
+    .map((token, index) => {
+      const previousBalance = response.transactions[index].return;
+
+      const newBalance =
+        response.transactions[index + balanceTransactions.length + 1].return;
+
+      const balanceChange = getSimulationBalanceChange(
+        previousBalance,
+        newBalance,
+      );
+
+      if (!balanceChange) {
+        return undefined;
+      }
+
+      return {
+        ...token,
+        ...balanceChange,
+      };
+    })
+    .filter((change) => change !== undefined) as SimulationTokenBalanceChange[];
+}
+
+/**
+ * Generate transactions to check token balances.
+ * @param request - The transaction that was simulated.
+ * @param events - The parsed events.
+ * @returns A map of token balance transactions keyed by token.
+ */
+function getTokenBalanceTransactions(
+  request: GetSimulationDataRequest,
+  events: ParsedEvent[],
+): Map<SimulationToken, SimulationRequestTransaction> {
+  const tokenKeys = new Set();
+
+  return events.reduce((result, event) => {
+    if (
+      !['Transfer', 'TransferSingle', 'TransferBatch'].includes(event.name) ||
+      ![event.data.from, event.data.to].includes(request.from)
+    ) {
+      log('Ignoring event', event);
+      return result;
+    }
+
+    let tokenIds: (Hex | undefined)[] = [undefined];
```

Comment:
It looks like `[undefined]` is used to initialize the ERC20 case, which doesn't have a tokenId, because the for-loop still needs an array with one element. Maybe add a comment to clarify this? Alternatively, you could make the ERC20 case more explicit:

```
let tokenIds: (Hex | undefined)[] = [];

if (event.tokenStandard === SimulationTokenStandard.erc20) {
    // A single tokenId of undefined for the for-loop below.
    tokenIds = [undefined]; 
}
```

Reply:
Done. Added a comment.